File: sdks/go/features/webhooks.go
package features

import (
	"context"

	"github.com/cockroachdb/errors"
	"github.com/google/uuid"

	"github.com/hatchet-dev/hatchet/pkg/client/rest"
)

type WebhookAuth interface {
	toCreateRequest(opts CreateWebhookOpts) (rest.V1CreateWebhookRequest, error)
}

type BasicAuth struct {
	Username string
	Password string
}

func (a BasicAuth) toCreateRequest(opts CreateWebhookOpts) (rest.V1CreateWebhookRequest, error) {
	var req rest.V1CreateWebhookRequest
	err := req.FromV1CreateWebhookRequestBasicAuth(rest.V1CreateWebhookRequestBasicAuth{
		Name:               opts.Name,
		SourceName:         opts.SourceName,
		EventKeyExpression: opts.EventKeyExpression,
		AuthType:           rest.V1CreateWebhookRequestBasicAuthAuthType("BASIC"),
		Auth: rest.V1WebhookBasicAuth{
			Username: a.Username,
			Password: a.Password,
		},
	})
	return req, err
}

type APIKeyAuth struct {
	HeaderName string
	APIKey     string
}

func (a APIKeyAuth) toCreateRequest(opts CreateWebhookOpts) (rest.V1CreateWebhookRequest, error) {
	var req rest.V1CreateWebhookRequest
	err := req.FromV1CreateWebhookRequestAPIKey(rest.V1CreateWebhookRequestAPIKey{
		Name:               opts.Name,
		SourceName:         opts.SourceName,
		EventKeyExpression: opts.EventKeyExpression,
		AuthType:           rest.V1CreateWebhookRequestAPIKeyAuthType("API_KEY"),
		Auth: rest.V1WebhookAPIKeyAuth{
			HeaderName: a.HeaderName,
			ApiKey:     a.APIKey,
		},
	})
	return req, err
}

type HMACAuth struct {
	SigningSecret       string
	SignatureHeaderName string
	Algorithm           rest.V1WebhookHMACAlgorithm
	Encoding            rest.V1WebhookHMACEncoding
}

func (a HMACAuth) toCreateRequest(opts CreateWebhookOpts) (rest.V1CreateWebhookRequest, error) {
	var req rest.V1CreateWebhookRequest
	err := req.FromV1CreateWebhookRequestHMAC(rest.V1CreateWebhookRequestHMAC{
		Name:               opts.Name,
		SourceName:         opts.SourceName,
		EventKeyExpression: opts.EventKeyExpression,
		AuthType:           rest.V1CreateWebhookRequestHMACAuthType("HMAC"),
		Auth: rest.V1WebhookHMACAuth{
			SigningSecret:       a.SigningSecret,
			SignatureHeaderName: a.SignatureHeaderName,
			Algorithm:           a.Algorithm,
			Encoding:            a.Encoding,
		},
	})
	return req, err
}

type CreateWebhookOpts struct {
	Name               string
	SourceName         rest.V1WebhookSourceName
	EventKeyExpression string
	Auth               WebhookAuth
}

type UpdateWebhookOpts struct {
	EventKeyExpression string
}

// WebhooksClient provides methods for managing webhook configurations
type WebhooksClient struct {
	api      *rest.ClientWithResponses
	tenantId uuid.UUID
}

// NewWebhooksClient creates a new WebhooksClient
func NewWebhooksClient(
	api *rest.ClientWithResponses,
	tenantId uuid.UUID,
) *WebhooksClient {
	tenantIdUUID := tenantId

	return &WebhooksClient{
		api:      api,
		tenantId: tenantIdUUID,
	}
}

// List retrieves a collection of webhooks based on the provided parameters.
func (c *WebhooksClient) List(ctx context.Context, opts rest.V1WebhookListParams) (*rest.V1WebhookList, error) {
	resp, err := c.api.V1WebhookListWithResponse(
		ctx,
		c.tenantId,
		&opts,
	)
	if err != nil {
		return nil, errors.Wrap(err, "failed to list webhooks")
	}

	if err := validateJSON200Response(resp.StatusCode(), resp.Body, resp.JSON200); err != nil {
		return nil, err
	}

	return resp.JSON200, nil
}

// Get retrieves a specific webhook by its name.
func (c *WebhooksClient) Get(ctx context.Context, webhookName string) (*rest.V1Webhook, error) {
	resp, err := c.api.V1WebhookGetWithResponse(
		ctx,
		c.tenantId,
		webhookName,
	)
	if err != nil {
		return nil, errors.Wrap(err, "failed to get webhook")
	}

	if err := validateJSON200Response(resp.StatusCode(), resp.Body, resp.JSON200); err != nil {
		return nil, err
	}

	return resp.JSON200, nil
}

// Create creates a new webhook configuration.
func (c *WebhooksClient) Create(ctx context.Context, opts CreateWebhookOpts) (*rest.V1Webhook, error) {
	if opts.Auth == nil {
		return nil, errors.New("auth is required")
	}

	req, err := opts.Auth.toCreateRequest(opts)
	if err != nil {
		return nil, errors.Wrap(err, "failed to create webhook request")
	}

	resp, err := c.api.V1WebhookCreateWithResponse(
		ctx,
		c.tenantId,
		req,
	)
	if err != nil {
		return nil, errors.Wrap(err, "failed to create webhook")
	}

	if err := validateJSON200Response(resp.StatusCode(), resp.Body, resp.JSON200); err != nil {
		return nil, err
	}

	return resp.JSON200, nil
}

// Update updates an existing webhook configuration.
func (c *WebhooksClient) Update(ctx context.Context, webhookName string, opts UpdateWebhookOpts) (*rest.V1Webhook, error) {
	resp, err := c.api.V1WebhookUpdateWithResponse(
		ctx,
		c.tenantId,
		webhookName,
		rest.V1UpdateWebhookRequest{
			EventKeyExpression: opts.EventKeyExpression,
		},
	)
	if err != nil {
		return nil, errors.Wrap(err, "failed to update webhook")
	}

	if err := validateJSON200Response(resp.StatusCode(), resp.Body, resp.JSON200); err != nil {
		return nil, err
	}

	return resp.JSON200, nil
}

// Delete removes a webhook configuration.
func (c *WebhooksClient) Delete(ctx context.Context, webhookName string) error {
	resp, err := c.api.V1WebhookDeleteWithResponse(
		ctx,
		c.tenantId,
		webhookName,
	)
	if err != nil {
		return errors.Wrap(err, "failed to delete webhook")
	}

	if err := validateStatusCodeResponse(resp.StatusCode(), resp.Body); err != nil {
		return err
	}

	return nil
}

File: sdks/typescript/src/v1/client/features/webhooks.ts
import {
  V1CreateWebhookRequest,
  V1CreateWebhookRequestBase,
  V1UpdateWebhookRequest,
  V1Webhook,
  V1WebhookAuthType,
  V1WebhookList,
  V1WebhookSourceName,
  V1WebhookAPIKeyAuth,
  V1WebhookBasicAuth,
  V1WebhookHMACAuth,
} from '@hatchet/clients/rest/generated/data-contracts';
import { HatchetClient } from '../client';

export type CreateWebhookOptions = V1CreateWebhookRequestBase & {
  auth: V1WebhookBasicAuth | V1WebhookAPIKeyAuth | V1WebhookHMACAuth;
};

function getAuthType(
  auth: V1WebhookBasicAuth | V1WebhookAPIKeyAuth | V1WebhookHMACAuth
): V1WebhookAuthType {
  if ('username' in auth && 'password' in auth) return V1WebhookAuthType.BASIC;
  if ('headerName' in auth && 'apiKey' in auth) return V1WebhookAuthType.API_KEY;
  if (
    'signingSecret' in auth &&
    'signatureHeaderName' in auth &&
    'algorithm' in auth &&
    'encoding' in auth
  ) {
    return V1WebhookAuthType.HMAC;
  }
  throw new Error('Invalid webhook auth');
}

function toCreateWebhookRequest(options: CreateWebhookOptions): V1CreateWebhookRequest {
  const { auth, ...base } = options;
  const authType = getAuthType(auth);
  return { ...base, authType, auth } as V1CreateWebhookRequest;
}

/**
 * Client for managing incoming webhooks in Hatchet.
 *
 * Webhooks allow external systems to trigger Hatchet workflows by sending HTTP
 * requests to dedicated endpoints. This enables real-time integration with
 * third-party services like GitHub, Stripe, Slack, or any system that can send
 * webhook events.
 */
export class WebhooksClient {
  api: HatchetClient['api'];
  tenantId: string;

  constructor(client: HatchetClient) {
    this.api = client.api;
    this.tenantId = client.tenantId;
  }

  async list(options?: {
    limit?: number;
    offset?: number;
    webhookNames?: string[];
    sourceNames?: V1WebhookSourceName[];
  }): Promise<V1WebhookList> {
    const response = await this.api.v1WebhookList(this.tenantId, {
      limit: options?.limit,
      offset: options?.offset,
      webhookNames: options?.webhookNames,
      sourceNames: options?.sourceNames,
    });
    return response.data;
  }

  async get(webhookName: string): Promise<V1Webhook> {
    const response = await this.api.v1WebhookGet(this.tenantId, webhookName);
    return response.data;
  }

  async create(request: CreateWebhookOptions): Promise<V1Webhook> {
    const payload = toCreateWebhookRequest(request);
    const response = await this.api.v1WebhookCreate(this.tenantId, payload);
    return response.data;
  }

  async update(
    webhookName: string,
    options: Partial<V1UpdateWebhookRequest> = {}
  ): Promise<V1Webhook> {
    const response = await this.api.v1WebhookUpdate(this.tenantId, webhookName, {
      eventKeyExpression: options.eventKeyExpression,
      scopeExpression: options.scopeExpression,
      staticPayload: options.staticPayload,
    });
    return response.data;
  }

  async delete(webhookName: string): Promise<V1Webhook> {
    const response = await this.api.v1WebhookDelete(this.tenantId, webhookName);
    return response.data;
  }
}
